lib/pos_mempool_nonce_tracker.go

```diff
@@ -0,0 +1,52 @@
+package lib
+
+// PosMempoolNonceTracker is a helper struct that is used to track (public key, nonce) pairs in pos mempool.
+// It is used to facilitate the "replace by higher fee" feature, which allows users to resubmit transactions.
+// If a user submits a transaction with an existing DeSoNonce.PartialID (among this user's transactions in the mempool),
+// then the new transaction will replace the old one if the new transaction has a higher fee.
+type PosMempoolNonceTracker struct {
+	// nonceMap indexes mempool transactions by (public key, nonce) pairs.
+	nonceMap map[PublicKey]map[uint64]*MempoolTx
+}
+
+func NewPosMempoolNonceTracker() *PosMempoolNonceTracker {
+	return &PosMempoolNonceTracker{
+		nonceMap: make(map[PublicKey]map[uint64]*MempoolTx),
+	}
+}
+
+// GetTxnByPublicKeyNonce returns the transaction with the given public key and nonce pair.
+func (pmnt *PosMempoolNonceTracker) GetTxnByPublicKeyNonce(pk PublicKey, nonce uint64) *MempoolTx {
+	innerMap, ok := pmnt.nonceMap[pk]
```

Comment:
Do we need a read lock here?

Reply:
IMO, it's fine to define this tracker with no internal mutex. As long as the parent PosMempool struct wraps the usage of the tracker with its own mutex, it's all kosher and simple to read.